src/meta/src/model/mod.rs

```diff
@@ -236,34 +238,207 @@ where
+}
 
-    /// For keys only in `self.orig_value`, call `delete_in_transaction` for the corresponding
-    /// value. For keys only in `self.new_value` or in both `self.new_value` and
-    /// `self.orig_value` but different in the value, call `upsert_in_transaction` for the
-    /// corresponding value.
-    fn apply_to_txn(&self, txn: &mut Transaction) -> Result<()> {
-        if let Some(new_value) = &self.new_value {
-            for (k, v) in self.orig_value_ref.iter() {
-                if !new_value.contains_key(k) {
-                    v.delete_in_transaction(txn)?;
+impl<'a, K: Ord, V: Clone> DerefMut for BTreeMapTransactionValueGuard<'a, K, V> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        let is_occupied = matches!(
+            self.staging_entry.as_ref().unwrap(),
+            BTreeMapTransactionStagingEntry::Occupied(_)
+        );
+
+        if !is_occupied {
+            let vacant_entry = match self.staging_entry.take().unwrap() {
+                BTreeMapTransactionStagingEntry::Vacant(entry) => entry,
+                BTreeMapTransactionStagingEntry::Occupied(_) => {
+                    unreachable!("we have previously check that the entry is not occupied")
                 }
+            };
+            let new_value_mut_ref = match vacant_entry.insert(BTreeMapOp::Insert(
+                self.orig_value
+                    .expect("self.staging_entry was vacant, so orig_value must be some")
+                    .clone(),
+            )) {
+                BTreeMapOp::Insert(v) => v,
+                BTreeMapOp::Delete => {
+                    unreachable!(
+                        "the previous inserted op is `Inserted`, so it's not possible to reach Delete"
+                    )
+                }
+            };
+            let _ = self
+                .staging_entry
+                .insert(BTreeMapTransactionStagingEntry::Occupied(new_value_mut_ref));
+        }
+
+        match self.staging_entry.as_mut().unwrap() {
+            BTreeMapTransactionStagingEntry::Vacant(_) => {
+                unreachable!("we have inserted a cloned original value in case of vacant")
             }
-            for (k, v) in new_value {
-                let orig_value = self.orig_value_ref.get(k);
-                if orig_value.is_none() || orig_value.unwrap() != v {
-                    v.upsert_in_transaction(txn)?;
+            BTreeMapTransactionStagingEntry::Occupied(v) => *v,
+        }
+    }
+}
+
+enum BTreeMapOp<V> {
+    Insert(V),
+    Delete,
+}
+
+/// A `ValTransaction` that wraps a `BTreeMap`. It supports basic `BTreeMap` operations like `get`,
+/// `get_mut`, `insert` and `remove`. Incremental modification of `insert`, `remove` and `get_mut`
+/// are stored in `staging`. On `commit`, it will apply the changes stored in `staging` to the in
+/// memory btree map. When serve `get` and `get_mut`, it merges the value stored in `staging` and
+/// `tree_ref`.
+pub struct BTreeMapTransaction<'a, K: Ord, V> {
```

Comment:
A big portion of the `BTreeMapTransaction` and `BTreeMapTransactionValueGuard` implementation is to deal with merging "staging" entry with "original" entry. Do we actually have a use case that we need to iter/read the BTreeMap after writing "staging" entry to it?